### ISSUE TITLE
[GFX950][MFMA]: Int8 dot generates wrong results when BLOCK_K % 4 != 0

### ISSUE THREAD

**Issue by pabloantoniom** [ROCmSoftwarePlatform/triton#958]
### Problem Description

An i8 dot whose K dimension is not a multiple of 4 cannot use MFMA (i8 MFMA needs kDim >= 32) and cannot use the integer FMA path either. See the guard in third_party/amd/lib/TritonAMDGPUTransforms/AccelerateAMDMatmul.cpp:

    // Try I8 x I8 -> I32 v_dot
    // if k % 4 != 0: can not use integer V_DOT instruction
    if (dotTypes.a.isInteger(8) && ... && k % 4 == 0)
        return true;

When that guard fails, `tryLegalizeFMA` promotes the whole dot to a floating type, including the loop-carried i32 accumulator, which is converted i32 -> f32 -> i32 on every iteration of the K loop. This results in precision loss, thus giving wrong results.

### Operating System

Ubuntu 22.04.5 LTS (Jammy Jellyfish)

### CPU

AMD Ryzen Threadripper PRO 7985WX

### GPU

AMD Instinct MI350X

### ROCm Version

ROCm 7

### Steps to Reproduce

Python reproducer:

```
#!/usr/bin/env python3

import numpy as np
import torch
import triton
import triton.language as tl

M = N = 256
K = 4608
BLOCK_M = 64
BLOCK_N = 32


@triton.jit
def matmul_i8_kernel(
    a_ptr, b_ptr, c_ptr,
    M, N, K,
    stride_am, stride_ak,
    stride_bk, stride_bn,
    stride_cm, stride_cn,
    BLOCK_M: tl.constexpr,
    BLOCK_N: tl.constexpr,
    BLOCK_K: tl.constexpr,
):
    pid_m = tl.program_id(0)
    pid_n = tl.program_id(1)

    offs_m = pid_m * BLOCK_M + tl.arange(0, BLOCK_M)
    offs_n = pid_n * BLOCK_N + tl.arange(0, BLOCK_N)
    offs_k = tl.arange(0, BLOCK_K)

    a_ptrs = a_ptr + offs_m[:, None] * stride_am + offs_k[None, :] * stride_ak
    b_ptrs = b_ptr + offs_k[:, None] * stride_bk + offs_n[None, :] * stride_bn

    acc = tl.zeros((BLOCK_M, BLOCK_N), dtype=tl.int32)
    for _ in range(0, K, BLOCK_K):
        a = tl.load(a_ptrs)
        b = tl.load(b_ptrs)
        acc = tl.dot(a, b, acc, out_dtype=tl.int32)
        a_ptrs += BLOCK_K * stride_ak
        b_ptrs += BLOCK_K * stride_bk

    c_ptrs = c_ptr + offs_m[:, None] * stride_cm + offs_n[None, :] * stride_cn
    tl.store(c_ptrs, acc)


rng = np.random.default_rng(0)
a_np = rng.integers(120, 128, size=(M, K), dtype=np.int8)
b_np = rng.integers(120, 128, size=(K, N), dtype=np.int8)
ref = a_np.astype(np.int64) @ b_np.astype(np.int64)

a = torch.from_numpy(a_np).cuda()
b = torch.from_numpy(b_np).cuda()

print(f"M={M} N={N} K={K} BLOCK_M={BLOCK_M} BLOCK_N={BLOCK_N}")
print(f"reference max element = {ref.max()} (f32 exact-integer limit = {1 << 24})\n")

for BLOCK_K in (2, 4):
    c = torch.empty((M, N), dtype=torch.int32, device=a.device)
    matmul_i8_kernel[(triton.cdiv(M, BLOCK_M), triton.cdiv(N, BLOCK_N))](
        a, b, c,
        M, N, K,
        a.stride(0), a.stride(1),
        b.stride(0), b.stride(1),
        c.stride(0), c.stride(1),
        BLOCK_M=BLOCK_M, BLOCK_N=BLOCK_N, BLOCK_K=BLOCK_K,
    )

    diff = np.abs(c.cpu().numpy().astype(np.int64) - ref)
    n_bad = int((diff != 0).sum())
    status = "OK" if n_bad == 0 else "MISMATCH"
    print(f"BLOCK_K={BLOCK_K}  {status:<9} "
          f"mismatched={n_bad}/{ref.size}  maxAbsDiff={int(diff.max())}")
```

Result:

```
$ python3 triton_i8_smallk_repro.py
M=256 N=256 K=4608 BLOCK_M=64 BLOCK_N=32
reference max element = 70405061 (f32 exact-integer limit = 16777216)

BLOCK_K=2  MISMATCH  mismatched=65127/65536  maxAbsDiff=301
BLOCK_K=4  OK        mismatched=0/65536  maxAbsDiff=0
```

### (Optional for Linux users) Output of /opt/rocm/bin/rocminfo --support

<details>
<summary>rocminfo --support output</summary>

```
Paste output here
```

</details>
